### PR TITLE
Add `Category` loading from YAML source with validation

### PR DESCRIPTION
**Pull request by danielpgross**
Similar to #443, but for categories: this PR adds the ability to load `Category` objects from a YAML source with validation.

This implementation **fully supports secondary paths** to a given category defined using a `secondary_children` key in the YAML source containing an array. Example:
```
- id: aa
  name: Apparel & Accessories
  children:
  - aa-1
  attributes:
  - color
- id: aa-1
  name: Clothing
  children: []
  secondary_children:
  - bi-19-10
- id: bi
  name: Business & Industrial
  children:
  - bi-19
  attributes:
  - color
- id: bi-19
  name: Medical
  children:
  - bi-19-10
  attributes:
  - color
- id: bi-19-10
  name: Scrubs
  children: []
  attributes:
  - color
```

The approach to loading categories is also different from the prototype: `load_from_source` takes the parsed YAML objects for _all_ verticals (not just one). This was needed because:

- We need to be able to resolve secondary paths, which can run through a different vertical
- We need to check uniqueness across all verticals

### CODE REVIEW

**Comment by danielpgross** [comment]
* **#449** <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/449?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/449?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: dev/lib/product_taxonomy/models/category.rb [hunk 1]
```diff
@@ -2,15 +2,246 @@
...
+    # @param [Category] category
+    # @return [Boolean]
+    def descendant_of?(category)
+      ancestors.include?(category)
+    end
+
+    # Iterate over the category and all its descendants
+    #
+    # @yield [Category]
+    def traverse(&block)
```
Comment (elsom25) [inline dev/lib/product_taxonomy/models/category.rb:178]: I assume usage of this will come in later PRs?

**Review by elsom25 — APPROVED** [review]
🙌

File: dev/lib/product_taxonomy/models/category.rb (hunk 1 above)
Comment (danielpgross) [reply dev/lib/product_taxonomy/models/category.rb:178]: For now I've been using it for debugging, ex. for recursively printing out the taxonomy

**Comment by danielpgross** [comment]
### Merge activity

* **Nov 19, 10:12 AM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Shopify/product-taxonomy/449).

**Pull request merged**